Support multi/diff addresses for noobaa services including auto discovery in k8s environments.

nb-ohad opened the pull request:
### Explain the changes
1. Refactor and rearrange create_system code
2. Support multi address for noobaa services including address discovery code for k8s environments 
3. Add support for RPC routing negotiation 
4. Add support for endpoint virtual hosts
5. Fix generation of 2 agent conf with the same name (it was time based and I anded a random suffix)

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1.

nb-ohad committed:
Support multi/diff addresses for noobaa services (+auto discovery in k8s envs)

- Refactor and rearrange create_system code
- Support multi address for noobaa services including address discovery code for k8s environments
- Add support for RPC routing negotiation and endpoint virtual hosts
- Fix generation of 2 agent conf with the same name
- Add routing hint to install conf to allow the initial connection from an agent to forward the correct hint.
- Add missing roles to noobaa yaml
- Add minimum address discovery for virtual appliance env so scripts who knows the public ip (for example azure.js) will continue
  to work

dannyzaken commented on src/util/os_utils.js:
```diff
@@ -1280,14 +1280,52 @@ async function is_vmtools_installed() {
 
-
-async function get_kubernetes_dns_name() {
+async function discover_k8s_services(app = 'noobaa') {
```
I think this function creates a strong coupling between the YAML files and k8s internals with our code:
1. using the `app=noobaa` label forces us to not change it in the future
2. using the port names in the yaml file forces us to not change it in the future
3. in general, I think that using kubernetes internals to get configuration data is not a good idea. any change in the API can break an existing noobaa deployment 

I suggest using [configMaps](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/) to pass configuration data, which are the native kubernetes tool for this job.  
For starter, we can manage the configMap internally in noobaa-server but eventually, I think the right approach is to update this data externally from the server to reduce dependencies.

nimrod-becker replied:
@nb-ohad Please open a tech-debt issue and attach it. Describe that we want to be able to retrieve this information from an outside source and use it.

nb-ohad replied:
All dependency on k8s in this code path is relegated to one function in os_utils.js (discover_k8s_services) so future changes or additions of  new discovery methods will be relatively easy

nb-ohad replied:
@nimrod-becker I don't know if this is what we want. You cannot total deprive the system from the knowledge that it is running on k8s (not only in this parts of the codebase but as a whole).
I think that converging all decisions based on env should be grouped into one area of the code base and specialised for each target env/distro and that there is not "evil' in doing so.

guymguym commented on src/server/bg_services/server_monitor.js:
```diff
@@ -503,5 +534,23 @@ async function _verify_and_update_public_ip() {
+    try {
+        const system_id = system_store.data.systems[0]._id;
+        const system_address = await os_utils.discover_k8s_services();
+        await system_store.make_changes({
```
Why do you need to call this make_changes() unconditionally?
The thing is that make_changes also needs to publish the changes and if there is no change to make we better avoid calling it.
Can you check if you really need to update?

nb-ohad commented on src/endpoint/endpoint.js:
```diff
@@ -114,7 +131,25 @@ function start_all() {
         const system_store = require('../server/system_services/system_store').get_instance(); // eslint-disable-line global-require
-        system_store.on('load', () => update_virtual_host_suffix(system_store.data.systems[0] && system_store.data.systems[0].base_address));
+        system_store.on('load', () => {
+            const [system] = system_store.data.systems;
```
@guymguym I don't see the diff between this syntax and object destructuring syntax which is wildly used in the codebase.

guymguym replied:
Object destructoring is name based, array destructoring is position based.
This is the difference. Specifically for position 0.

nb-ohad commented on src/endpoint/s3/s3_rest.js:
```diff
@@ -231,63 +231,70 @@ function authenticate_request(req, res) {
+    } else {
+        const match = virtual_hosts
+            .map(virtual_host => `.${virtual_host}`)
+            .find(suffix => host.endsWith(suffix));
```
@guymguym Array.map and Array.find are not actual calls to some native/internal functions in V8. 
The actual generated V8 op code will remove and inline both this function into something more in the like of the loop you describe. 
On the other hand for of loop are more expansive because of the iterator result that is generated for every item (maybe in the future the V8 theme will optimize the "for of" for simple arrays to skip the iterator all together).

Regarding the code readability, I always prefer to code more functional than imperative over arrays (map, filter, reduce, fund, etc.) because you don't need to specify how to loop but you want to achieve and in my eyes it is even more readable.

saying all that the agree that the map is redundant and i will change it to:
```
const match = virtual_hosts.find(virtual_host => host.endsWith(`.${virtual_host}`));
```
which keep match a const and not a var (or a let), this will give the compiler more opportunities for optimizations, like not allocating a frame variable the const

guymguym replied:
ok

nb-ohad commented on src/endpoint/s3/s3_rest.js:
```diff
@@ -231,63 +231,70 @@ function authenticate_request(req, res) {
+
+    let { bucket, key, is_virtual_host_bucket } = parse_bucket_and_key(req);
     req.params = { bucket, key };
+    req.virtual_host_bucket = is_virtual_host_bucket;
```
The bucket is already returned in the bucket params (also the key in the key param).
This is only and indication that the bucket is a virtual host bucket.
This was the interface of the result of this function before and i didn't change it

guymguym replied:
Do you see the problem though? First a typo, second that req.virtual_hosted_bucket should be the bucket name for signature_utils to work properly.

guymguym replied:
As we discussed, this affects only Signature v2 so it wasn't noticed in the test.
We want to add a tests on virt host which covers both sig v4 and v2.

nb-ohad commented on src/endpoint/endpoint.js:
```diff
@@ -43,11 +43,11 @@ const { FtpFileSystemNB } = require('./ftp/ftp_filesystem');
 // or when a system store load was done if server
-let virtual_host_suffix;
-// this "shared" object is used to save the location information on s3 agents 
+let virtual_hosts = [];
```
we cannot freeze it because it is being updated at line 384

guymguym replied:
The reference is updated, not the object. So it can be frozen, right?

nb-ohad commented on src/util/url_utils.js:
```diff
@@ -50,6 +46,34 @@ function quick_parse(url_string, parse_query_string) {
+        throw new Error('Invalid definition, hostname is mandatory');
+    }
+
+    const u = new URL(`${def.protocol || 'http'}://${def.hostname}`);
```
@guymguym Sorry but I don't see the "parse", Can you please clarify?
You suggestion is also using the URL constructor with a string. what is the difference?

nb-ohad commented on src/endpoint/endpoint.js:
```diff
@@ -114,7 +131,25 @@ function start_all() {
+                .map(addr => addr.hostname);
+
+            const virtual_hosts_list = _.uniq(matching_hostnames).sort();
+            update_virtual_hosts(virtual_hosts_list);
```
This is not the only use of update_virtual_hosts, see code at lines 92 and 96 which cover the path where the endpoint is embedded in an agent.

In order to make the code more readable I now moved this code patch into it's own function:
```update_virtual_hosts_form_system```

nb-ohad commented on src/endpoint/endpoint.js:
```diff
@@ -337,16 +374,15 @@ function setup_http_server(server) {
+function update_virtual_hosts(new_virtual_hosts) {
+    new_virtual_hosts = new_virtual_hosts
+        .filter(suffix => net_utils.is_fqdn(suffix))
+        .sort();
```
Removed the sort from the user code (system_store.on('load', () => ...) left it here. 
The filter still needed because the list can come for a diff source (not only the system_store list)

nb-ohad commented on src/endpoint/s3/s3_rest.js:
```diff
@@ -231,63 +231,70 @@ function authenticate_request(req, res) {
-        host.endsWith(virtual_host_suffix) &&
-        host !== virtual_host_suffix; // non empty bucket name
+        host === 'localhost' || // we added this case on top of the S3 doc cases to handle requests with IP host
+        virtual_hosts.length === 0 ||
```
virtual_hosts is always set, first to an empty array then to the array given from the agent or loaded from the system_store (depending if the endpoint is embedded in the server or an agent) inside ```update_virtual_hosts(new_virtual_hosts)```

guymguym replied:
that's true for the global, not for `req.virtual_hosts` which is the one you are accessing here

nb-ohad replied:
@guymguym the virtual_hosts on the request was set in  endpoint.js in ```endpoint_request_handler```. |
 We just attach to the req the global array of virtual hosts that the endpoint is holding which as i explained above is always set to something (an empty array or an array with hosts names) after the new commit this array is also frozen

guymguym replied:
ok

nb-ohad commented on src/endpoint/endpoint.js:
```diff
@@ -217,7 +252,9 @@ function create_endpoint_handler(rpc, options) {
         }
 
-        if (virtual_host_suffix) req.virtual_host_suffix = virtual_host_suffix;
+        if (virtual_hosts) {
```
Is is already always set, the if is redundant. I will remove it

nb-ohad commented on src/util/os_utils.js:
```diff
@@ -1280,14 +1280,52 @@ async function is_vmtools_installed() {
+                    hostname
+                }))
+            ];
+        }).sort((item, other) =>
```
I am sorting by a using a field of an object (the "kind" field) os default sorting will not do

nb-ohad commented:
To anyone who is reviewing, for some reason commits from the master got merged into the history of this branch, so please ignor. I will fix this tomorrow

guymguym commented:
@nb-ohad Can you resolve the reviews that you fixed? So that we can see the status by what's left unresolved...

guymguym commented on src/test/unit_tests/test_system_servers.js:
```diff
@@ -129,9 +129,10 @@ mocha.describe('system_servers', function() {
-            .then(() => rpc_client.system.update_base_address({
-                base_address: 'fcall://fcall'
-            }))
+            // TODO: underastand what is happning here
```
unneeded.

dannyzaken commented:
@nb-ohad @nimrod-becker is there a new base yet?

dannyzaken commented on src/agent/agent.js:
```diff
@@ -733,6 +734,8 @@ class Agent {
                     base_address: this.master_address,
                     certs: this._ssl_certs,
                     location_info: this.location_info,
+                    routing_hint: this.rpc.routing_hint,
```
if I understand correctly the routing hint is what the agent thinks its location is (internal\external). what about the case of a cluster where we can have multiple noobaa server that each can be in a differnet location. do we know what to do when we'll get there?

nb-ohad replied:
Currently we do not support this use case in the full. 
The current configuration is that agents are internal to the k8s that they ware installed in (and external to anybody else). In the future each location will get its own tag (instead of INTERNAL and EXTERNAL - the schema supports it) and the routing table will be calculated and send back to the client (the agent/endpoint) taking into account the client tag and the routing authority tag.

dannyzaken commented on src/endpoint/s3/s3_rest.js:
```diff
@@ -231,63 +231,65 @@ function authenticate_request(req, res) {
-
-    if (is_path_style) {
+// We removed support for default website hosting (host value is a bucket name)
+// after discussion with eran and gay because it intreeduced a conflict.
```
did you mean guy? (not that there is anything wrong with that)

guymguym replied:
thanks @dannyzaken for the COC watch 🥇

guymguym commented on src/endpoint/s3/s3_rest.js:
```diff
@@ -231,63 +231,65 @@ function authenticate_request(req, res) {
-
-    if (is_path_style) {
+// We removed support for default website hosting (host value is a bucket name)
+// after discussion with eran and gay because it intreeduced a conflict.
```
```suggestion
// after a discussion with [Ohad, Nimrod, Eran, Guy] because it introduced a conflict
// that will be resolved with Bucket Website Configuration
```

nb-ohad committed:
Remove windows installation commands

We do not support windows agents anymore

nb-ohad committed:
Fixes issues

Allow system address changes to be initiated only from master
If not every member in the cluster will update it’s own address as the
public address of the system (they will overwrite each other)

Fix missing system_store.is_cluster_master for code which is not running in the web server

Enforce explicit set of a hint to api.new_rpc_from_base_address, removes
confusion regarding when a routing negotiation occurs

Optimize endpoint <-> md server access when both are deployed on the same machine

K8s - Add another location that exposes external names manifested in k8s on google cloud
See https://github.com/noobaa/noobaa-core/issues/5474

Add missing routing hint for hosted agents

nb-ohad committed:
Update src/endpoint/s3/s3_rest.js

Co-Authored-By: Guy Margalit <guymguym@gmail.com>
